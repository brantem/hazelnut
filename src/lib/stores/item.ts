import create from 'zustand';

type ItemState = {
  groupId: string | null;
  clear: () => void;

  isAddOpen: boolean;
  showAdd: (groupId: string) => void;

  hide: () => void;
};

export const useItemStore = create<ItemState>()((set) => ({
  groupId: null,
  clear: () => set({ groupId: null }),

  isAddOpen: false,
  showAdd: (groupId) => set({ groupId, isAddOpen: true }),

  hide: () => set({ isAddOpen: false }),
}));
