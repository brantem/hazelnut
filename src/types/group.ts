export type Group = {
  id: string;
  title: string;
  color: string;
  minimized: boolean;
  createdAt: number;
};
