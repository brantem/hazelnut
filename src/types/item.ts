export type Item = {
  id: string;
  groupId: string;
  title: string;
  createdAt: number;
};
