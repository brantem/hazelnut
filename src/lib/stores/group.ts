import create from 'zustand';

import { Group } from 'types/group';

type GroupState = {
  group: Group | null;
  clear: () => void;

  isSaveOpen: boolean;
  showSave: (group?: Group | null) => void;

  isSettingsOpen: boolean;
  showSettings: (group: Group) => void;

  hide: () => void;
};

export const useGroupStore = create<GroupState>()((set) => ({
  group: null,
  clear: () => set({ group: null }),

  isSaveOpen: false,
  showSave: (group = null) => set((state) => ({ group: group || state.group, isSaveOpen: true })),

  isSettingsOpen: false,
  showSettings: (group) => set({ group, isSettingsOpen: true }),

  hide: () => set({ isSaveOpen: false, isSettingsOpen: false }),
}));
