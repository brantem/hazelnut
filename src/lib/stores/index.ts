import { useEffect } from 'react';

import { itemsStore } from 'lib/stores/items';
import { groupsStore } from 'lib/stores/groups';
import { routinesStore } from 'lib/stores/routines';
import { historiesStore } from 'lib/stores/histories';
import storage from 'lib/stores/storage';

/* c8 ignore start */
export const useLoadStore = () => {
  useEffect(() => {
    (async () => {
      const [items, groups, routines, histories] = await Promise.all([
        storage.getAll('items'),
        storage.getAll('groups'),
        storage.getAll('routines'),
        storage.getAll('histories'),
      ]);

      itemsStore.setState({ items: items.sort((a, b) => a.createdAt - b.createdAt), isReady: true });
      groupsStore.setState({ groups: groups.sort((a, b) => a.createdAt - b.createdAt), isReady: true });
      routinesStore.setState({ routines, isReady: true });
      historiesStore.setState({ histories, selectedDate: localStorage.getItem('history-selected-date') });
    })();
  }, []);
};
/* c8 ignore stop */

export * from 'lib/stores/items';
export * from 'lib/stores/groups';
export * from 'lib/stores/routines';
export * from 'lib/stores/search';
export * from 'lib/stores/histories';
export * from 'lib/stores/modal';
