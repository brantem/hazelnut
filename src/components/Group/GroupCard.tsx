import { EllipsisHorizontalIcon } from '@heroicons/react/20/solid';
import { MinusCircleIcon } from '@heroicons/react/24/outline';
import clsx from 'clsx';

import DeleteButton from 'components/DeleteButton';

import { useGroupsStore, useItemsStore } from 'lib/stores';
import type { Group } from 'types/group';

type GroupCardProps = {
  group: Group;
};

const GroupCard = ({ group }: GroupCardProps) => {
  const { showSettings } = useGroupsStore();
  const items = useItemsStore((state) => state.items.filter((item) => item.groupId === group.id));
  const { showAdd, remove } = useItemsStore();

  return (
    <div className={`px-4 py-3 bg-${group.color}-50`} data-testid="group-card">
      <div className="flex items-center justify-between space-x-3">
        <h3 className={`text-sm font-semibold uppercase text-${group.color}-600 truncate`}>{group.title}</h3>

        <div className="flex flex-shrink-0 items-center space-x-1">
          <button
            className={`rounded-md px-2 py-1 text-sm hover:bg-${group.color}-100 flex-shrink-0`}
            onClick={() => showAdd(group.id)}
          >
            Add Item
          </button>

          <button
            className={`rounded-md p-1 hover:bg-${group.color}-100`}
            onClick={() => showSettings(group)}
            data-testid="group-card-settings"
          >
            <EllipsisHorizontalIcon className="h-5 w-5" />
          </button>
        </div>
      </div>

      {items.length ? (
        <ol className="space-y-1 pt-2 pb-1" data-testid="group-card-items">
          {items.map((item) => (
            <li
              data-testid="group-card-items-item"
              key={item.id}
              className="flex items-center justify-between space-x-3"
            >
              <span className="truncate">{item.title}</span>

              <DeleteButton
                className={(clicked) => clsx(`rounded-md p-1 text-sm text-red-500 hover:bg-red-100`, clicked && 'px-2')}
                text={<MinusCircleIcon className="h-5 w-5" />}
                onConfirm={() => remove(item.id)}
              />
            </li>
          ))}
        </ol>
      ) : null}
    </div>
  );
};

export default GroupCard;
