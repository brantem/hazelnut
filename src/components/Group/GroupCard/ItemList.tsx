import { useCallback } from 'react';
import { MinusCircleIcon } from '@heroicons/react/24/outline';
import clsx from 'clsx';

import DeleteButton from 'components/DeleteButton';

import { useItemsStore } from 'lib/stores';
import type { Group } from 'types/group';

type ItemListProps = {
  group: Group;
};

const ItemList = ({ group }: ItemListProps) => {
  const items = useItemsStore(useCallback((state) => state.getItemsByGroupId(group.id), [group.id]));
  const remove = useItemsStore((state) => state.remove);

  return (
    <ol className="space-y-1 pt-2 pb-1" data-testid="group-card-items">
      {items.map((item) => (
        <li data-testid="group-card-items-item" key={item.id} className="flex items-center justify-between space-x-3">
          <span className="truncate">{item.title}</span>

          <DeleteButton
            className={(clicked) => clsx(`rounded-md !p-1 text-sm text-red-500 hover:bg-red-100`, clicked && '!px-2')}
            text={<MinusCircleIcon className="h-5 w-5" />}
            onConfirm={() => remove(item.id)}
          />
        </li>
      ))}
    </ol>
  );
};

export default ItemList;
