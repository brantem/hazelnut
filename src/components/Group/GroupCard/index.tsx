import { EllipsisHorizontalIcon, PlusIcon, ChevronUpIcon } from '@heroicons/react/20/solid';

import clsx from 'clsx';

import ItemList from 'components/Group/GroupCard/ItemList';

import { useGroupsStore, useItemsStore } from 'lib/stores';
import type { Group } from 'types/group';

type GroupCardProps = {
  group: Group;
};

const GroupCard = ({ group }: GroupCardProps) => {
  const { showSettings, edit } = useGroupsStore();
  const { showAdd } = useItemsStore();

  return (
    <div className={`px-4 py-3 bg-${group.color}-50`} data-testid="group-card">
      <div className="flex items-center justify-between space-x-3">
        <h3 className={`text-sm font-semibold uppercase text-${group.color}-600 truncate`}>{group.title}</h3>

        <div className="flex flex-shrink-0 items-center space-x-1">
          <button
            className={`rounded-md p-1 text-sm hover:bg-${group.color}-100 flex-shrink-0`}
            onClick={() => showAdd(group.id)}
            data-testid="group-card-add-item"
          >
            <PlusIcon className="h-5 w-5" />
          </button>

          <button
            className={`rounded-md p-1 hover:bg-${group.color}-100`}
            onClick={() => showSettings(group)}
            data-testid="group-card-settings"
          >
            <EllipsisHorizontalIcon className="h-5 w-5" />
          </button>

          <button
            className={clsx(`rounded-md p-1 hover:bg-${group.color}-100`, group.minimized && 'rotate-180')}
            onClick={() => edit(group.id, { minimized: !group.minimized })}
            data-testid="group-card-minimize"
          >
            <ChevronUpIcon className="h-5 w-5" />
          </button>
        </div>
      </div>

      {!group.minimized && <ItemList group={group} />}
    </div>
  );
};

export default GroupCard;
