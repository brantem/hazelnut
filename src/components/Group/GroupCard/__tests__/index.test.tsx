import { render, screen, act, renderHook } from '@testing-library/react';
import '@testing-library/jest-dom';

import GroupCard from 'components/Group/GroupCard';

import { useGroupsStore, useItemsStore } from 'lib/stores';
import { Item } from 'types/item';
import { Group } from 'types/group';

const group: Group = {
  id: 'group-1',
  title: 'Group 1',
  color: 'red',
  minimized: false,
};

describe('GroupCard', () => {
  beforeEach(() => {
    const items = renderHook(() => useItemsStore());
    act(() => items.result.current.add('group-1', { id: 'item-1', title: 'Item 1' } as Item));
  });

  afterEach(() => {
    const items = renderHook(() => useItemsStore());
    act(() => items.result.current.remove('item-1'));
  });

  it('should render successfully', () => {
    const items = renderHook(() => useItemsStore());
    const showAdd = vi.spyOn(items.result.current, 'showAdd').mockImplementation(() => {});

    const { result } = renderHook(() => useGroupsStore());
    const showSettings = vi.spyOn(result.current, 'showSettings').mockImplementation(() => {});

    const { container } = render(<GroupCard group={group} />);

    expect(container).toMatchSnapshot();
    expect(screen.getAllByTestId('group-card-items-item')).toHaveLength(1);

    act(() => screen.getByTestId('group-card-add-item').click());
    expect(showAdd).toHaveBeenCalledWith('group-1');

    act(() => screen.getByTestId('group-card-settings').click());
    expect(showSettings).toHaveBeenCalledWith(group);
  });

  it('should remove item', () => {
    const items = renderHook(() => useItemsStore());
    const remove = vi.spyOn(items.result.current, 'remove');

    const { rerender } = render(<GroupCard group={group} />);

    expect(screen.getByTestId('group-card-items-item')).toBeInTheDocument();
    act(() => screen.getByTestId('delete-button').click());
    act(() => screen.getByTestId('delete-button-confirm').click());
    expect(remove).toHaveBeenCalledWith('item-1');
    rerender(<GroupCard group={group} />);
    expect(screen.queryByTestId('group-card-items-item')).not.toBeInTheDocument();
  });

  it('should be minimizable', () => {
    const { result } = renderHook(() => useGroupsStore());
    const edit = vi.spyOn(result.current, 'edit').mockImplementation(() => {});

    const { rerender } = render(<GroupCard group={group} />);

    expect(screen.getByTestId('group-card-items')).toBeInTheDocument();
    act(() => screen.getByTestId('group-card-minimize').click());
    expect(edit).toHaveBeenCalledWith('group-1', { minimized: true });
    rerender(<GroupCard group={{ ...group, minimized: true }} />);
    expect(screen.queryByTestId('group-card-items')).not.toBeInTheDocument();
  });
});
