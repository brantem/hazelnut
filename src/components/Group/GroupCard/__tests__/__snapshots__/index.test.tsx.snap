// Vitest Snapshot v1

exports[`GroupCard > should render successfully 1`] = `
<div>
  <div
    class="px-4 py-3 bg-red-50"
    data-testid="group-card"
  >
    <div
      class="flex items-center justify-between space-x-3"
    >
      <h3
        class="text-sm font-semibold uppercase text-red-600 truncate"
      >
        Group 1
      </h3>
      <div
        class="flex flex-shrink-0 items-center space-x-1"
      >
        <button
          class="rounded-md p-1 text-sm hover:bg-red-100 flex-shrink-0"
          data-testid="group-card-add-item"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10.75 4.75a.75.75 0 00-1.5 0v4.5h-4.5a.75.75 0 000 1.5h4.5v4.5a.75.75 0 001.5 0v-4.5h4.5a.75.75 0 000-1.5h-4.5v-4.5z"
            />
          </svg>
        </button>
        <button
          class="rounded-md p-1 hover:bg-red-100"
          data-testid="group-card-settings"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 10a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zM8.5 10a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zM15.5 8.5a1.5 1.5 0 100 3 1.5 1.5 0 000-3z"
            />
          </svg>
        </button>
        <button
          class="rounded-md p-1 hover:bg-red-100"
          data-testid="group-card-minimize"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="currentColor"
            viewBox="0 0 20 20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M14.77 12.79a.75.75 0 01-1.06-.02L10 8.832 6.29 12.77a.75.75 0 11-1.08-1.04l4.25-4.5a.75.75 0 011.08 0l4.25 4.5a.75.75 0 01-.02 1.06z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
      </div>
    </div>
    <ol
      class="space-y-1 pt-2 pb-1"
      data-testid="group-card-items"
    >
      <li
        class="flex items-center justify-between space-x-3"
        data-testid="group-card-items-item"
      >
        <span
          class="truncate"
        >
          Item 1
        </span>
        <button
          class="rounded-md p-1 text-sm text-red-500 hover:bg-red-100"
          data-testid="delete-button"
        >
          <svg
            aria-hidden="true"
            class="h-5 w-5"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 12H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </li>
    </ol>
  </div>
</div>
`;
